tests/late_joiners.py

```diff
@@ -104,107 +107,106 @@ def run_requests(nodes, total_requests, start_id, final_msg, final_msg_id):
-            # add new node to backups list
-            all_nodes.append(new_node)
+            nodes_to_kill = []
+            nodes_to_kill.append(backups[0])
+            nodes_to_keep = [first_node, backups[1], backups[2]]
 
             # check that a new node can catch up after all the requests
             LOG.info("Adding a very late joiner")
-            last_node = network.create_and_trust_node(
+            last_joiner_node = network.create_and_trust_node(
                 lib_name=args.package, host="localhost", args=args,
             )
-            assert last_node
-            nodes_to_keep.append(last_node)
+            assert last_joiner_node
+            nodes_to_keep.append(last_joiner_node)
 
-            run_requests(all_nodes, TOTAL_REQUESTS, 1001, second_msg, 2000)
+            # some requests to be processed simultaneously with the late joiner catching up
+            run_requests(all_nodes, int(TOTAL_REQUESTS / 4), 1001, second_msg, 2000)
             term_info.update(find_primary(network))
 
-            assert_node_up_to_date(check, last_node, first_msg, 1000)
-            assert_node_up_to_date(check, last_node, second_msg, 2000)
+            assert_node_up_to_date(check, last_joiner_node, first_msg, 1000)
+            assert_node_up_to_date(check, last_joiner_node, second_msg, 2000)
 
-            # replace the 2 backups with the 2 new nodes, kill the old ones and ensure we are still making progress
+            # kill the old node(s) and ensure we are still making progress with the new one(s)
             for node in nodes_to_kill:
                 LOG.info(f"Stopping node {node.node_id}")
                 node.stop()
 
-            wait_for_nodes(nodes_to_keep, final_msg, 4000)
+            wait_for_nodes(nodes_to_keep, catchup_msg, 3000)
 
-            # we have asserted that all nodes are caught up
+            # first timer determines after how many seconds each node will be suspended
+            timeouts = []
+            suspended_nodes = []
+            for i, node in enumerate(nodes_to_keep):
+                # if pbft suspend half of them including the primary
+                if i % 2 != 0 and args.consensus == "pbft":
+                    continue
+                LOG.success(f"Will suspend node with id {node.node_id}")
+                t = random.uniform(1, 10)
+                LOG.info(f"Initial timer for node {node.node_id} is {t} seconds...")
+                timeouts.append((t, node))
+                suspended_nodes.append(node.node_id)
 
-            if not args.skip_suspension:
-                # assert that view changes actually did occur
-                assert len(term_info) > 1
+            for t, node in timeouts:
+                et = args.election_timeout / 1000
+                # if pbft suspend the primary more than the other suspended nodes
+                if node.node_id == 0 and args.consensus == "pbft":
+                    et += 4
```

Comment:
Why 4? Would it better to just double `et`?

Reply:
That will suspend the node for too long, if it is set to be suspended for e.g. 20 seconds then it will be suspended for 40 seconds. We just want to make sure the primary sleeps a bit more than the other node that is suspended.

Reply:
Good point. Conversely, if `election-timeout` is set to say 1s, then the primary will sleep _a lot_ more than the other nodes. Could it be a % of the election timeout?

Reply:
That's probably better, I'll set it to `et += et * 0.5`